src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs

```diff
@@ -1283,28 +1351,40 @@ public TValue AddOrUpdate(TKey key, Func<TKey, TValue> addValueFactory, Func<TKe
                 ThrowHelper.ThrowArgumentNullException(nameof(updateValueFactory));
             }
 
-            IEqualityComparer<TKey>? comparer = _comparer;
-            int hashcode = comparer is null ? key.GetHashCode() : comparer.GetHashCode(key);
+            Tables tables = _tables;
+
+            IEqualityComparer<TKey>? comparer = tables._comparer;
+            int hashcode = GetHashCode(comparer, key);
 
             while (true)
             {
-                if (TryGetValueInternal(key, hashcode, out TValue? oldValue))
+                if (TryGetValueInternal(tables, key, hashcode, out TValue? oldValue))
                 {
                     // key exists, try to update
                     TValue newValue = updateValueFactory(key, oldValue);
-                    if (TryUpdateInternal(key, hashcode, newValue, oldValue))
+                    if (TryUpdateInternal(tables, key, hashcode, newValue, oldValue))
                     {
                         return newValue;
                     }
                 }
                 else
                 {
                     // key doesn't exist, try to add
-                    if (TryAddInternal(key, hashcode, addValueFactory(key), updateIfExists: false, acquireLock: true, out TValue resultingValue))
+                    if (TryAddInternal(tables, key, hashcode, addValueFactory(key), updateIfExists: false, acquireLock: true, out TValue resultingValue))
                     {
                         return resultingValue;
                     }
                 }
+
+                if (tables != _tables)
```

Comment:
I think it is bad idea to use AddOrUpdate with `IDisposable`. 

If we cache the result of `addValueFactory(key)` then it will likely hit `TryUpdate` branch in second iteration where we can:
- use cached result instead of `updateValueFactory(key, oldValue)` result and oldValue could not be possibly disposed in `updateValueFactory(key, oldValue)`.
- or discard the cached result anyway and use `updateValueFactory(key, oldValue)`.

It is unlikely to call `addValueFactory(key)` more than once.

Reply:
> It is unlikely to call addValueFactory(key) more than once.

Yeah, for this to occur, one thread would need to try the get and fail because the item wasn't there, then another thread would need to concurrently add by the time the first thread does its try add, which fails.  Then prior to the next get attempt another thread would need to remove, such that the get again fails.  It's certainly possible, which is why the retry loop exists, but it's not something we need to try to optimize for, and as was called out, it's a visible change in behavior that's not worth the risk at this point.